File: internal/limiter/distributed/leakybucket.go
package distributed

import (
	"context"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"
	"github.com/streamliner/rate-limiter/internal/limiter"
)

// LeakyBucket is a Redis-backed leaky bucket enforcing a constant output rate.
// This implementation handles rate computation atomically in Lua; the Go layer
// handles any actual request queuing/holding per node.
type LeakyBucket struct {
	spec  limiter.Spec
	rdb   *redis.Client
	clock limiter.Clock
}

func NewLeakyBucket(spec limiter.Spec, rdb *redis.Client, clock limiter.Clock) *LeakyBucket {
	return &LeakyBucket{spec: spec, rdb: rdb, clock: clock}
}

func (lb *LeakyBucket) Allow(ctx context.Context, key string, cost int64) (limiter.Decision, error) {
	if cost <= 0 {
		cost = 1
	}

	now := lb.clock.Now()
	nowMs := now.UnixMilli()
	capacity := float64(lb.spec.Capacity())
	drainPerSec := lb.spec.Rate()
	ttlMs := lb.spec.Window.Milliseconds() * 2

	redisKey := fmt.Sprintf("ratelimit:{%s}:lb", key)

	res, err := runScript(ctx, lb.rdb, leakyBucketScript,
		[]string{redisKey},
		capacity, drainPerSec, nowMs, float64(cost), ttlMs,
	)
	if err != nil {
		return limiter.Decision{}, err
	}

	allowed := toInt64(res[0])
	remaining := toInt64(res[1])
	retryMs := toInt64(res[2])

	retryAfter := time.Duration(retryMs) * time.Millisecond
	resetAfter := time.Duration(float64(remaining) / drainPerSec * float64(time.Second))

	if allowed == 1 {
		return limiter.Allowed(lb.spec.Limit, remaining, resetAfter), nil
	}
	return limiter.Denied(lb.spec.Limit, retryAfter, resetAfter), nil
}

File: internal/limiter/distributed/lua/leaky_bucket.lua
-- leaky_bucket.lua
-- Leaky bucket: constant drain rate, bursts absorbed up to capacity.
-- Rate computation only — request queuing/holding is the Go layer's responsibility.
--
-- KEYS[1]  = bucket key (HASH: fields "level" and "last_drain_ms")
-- ARGV[1]  = capacity       (float, max water level)
-- ARGV[2]  = drain_per_sec  (float, = limit / window_seconds)
-- ARGV[3]  = now_ms         (int, from Go — never call TIME here)
-- ARGV[4]  = cost           (float, water added by this request)
-- ARGV[5]  = ttl_ms         (int)
--
-- Returns: {allowed (0|1), remaining (int), retry_after_ms (int)}

local capacity      = tonumber(ARGV[1])
local drain_per_sec = tonumber(ARGV[2])
local now           = tonumber(ARGV[3])
local cost          = tonumber(ARGV[4])
local ttl           = tonumber(ARGV[5])

local state         = redis.call('HMGET', KEYS[1], 'level', 'last_drain_ms')
local level         = tonumber(state[1]) or 0
local last_drain_ms = tonumber(state[2]) or now

-- Drain the bucket based on elapsed time since last call.
local elapsed_s = math.max(0, (now - last_drain_ms) / 1000.0)
level = math.max(0, level - elapsed_s * drain_per_sec)

local retry_after_ms = 0

if level + cost <= capacity then
    level = level + cost
    redis.call('HSET', KEYS[1], 'level', level, 'last_drain_ms', now)
    redis.call('PEXPIRE', KEYS[1], ttl)

    local remaining = math.floor(capacity - level)
    return {1, remaining, 0}
end

-- Time until enough water drains to accept `cost` more units.
local overflow   = (level + cost) - capacity
retry_after_ms   = math.ceil(overflow / drain_per_sec * 1000)
local remaining  = math.floor(capacity - level)

redis.call('HSET', KEYS[1], 'level', level, 'last_drain_ms', now)
redis.call('PEXPIRE', KEYS[1], ttl)

return {0, remaining, retry_after_ms}

